### PR TITLE
Cleanup 'active' declare_features! with uniform style + sorting.

### PR DESCRIPTION
**Pull request by Centril**
r? @oli-obk 

(added the FIXME you wanted)
cc https://github.com/rust-lang/rust/pull/60354
cc https://github.com/rust-lang/rust/issues/60361

### CODE REVIEW

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -108,124 +108,227 @@ macro_rules! declare_features {
...
+    // macros disappear).
+    (active, allow_internal_unsafe, "1.0.0", None, None),
+
+    // Allows using the macros:
+    // + `__diagnostic_used`
+    // + `__register_diagnostic`
+    // +`__build_diagnostic_array`
     (active, rustc_diagnostic_macros, "1.0.0", None, None),
+
+    // Allows using `#[rustc_const_unstable (feature)]` which lets
```
Comment (oli-obk) [inline src/libsyntax/feature_gate.rs]:
```suggestion
    // Allows using `#[rustc_const_unstable(feature)]` which lets
```

File: src/libsyntax/feature_gate.rs [hunk 2]
```diff
@@ -108,124 +108,227 @@ macro_rules! declare_features {
...
+    // Allows defining identifiers beyond ASCII.
+    (active, non_ascii_idents, "1.0.0", Some(55467), None),
+
+    // Allows using `#[plugin_registrar]` on functions.
+    (active, plugin_registrar, "1.0.0", Some(29597), None),
+
+    // Allows using `#![plugin(myplugin)]`.
+    (active, plugin, "1.0.0", Some(29597), None),
+
+    // Allows using `#{thread_local]` on `static` items.
```
Comment (oli-obk) [inline src/libsyntax/feature_gate.rs]:
```suggestion
    // Allows using `#[thread_local]` on `static` items.
```

File: src/libsyntax/feature_gate.rs [hunk 3]
```diff
@@ -108,124 +108,227 @@ macro_rules! declare_features {
...
+    (active, abi_vectorcall, "1.7.0", None, None),
+
+    // no tracking issue END
+
     // Allows using the `may_dangle` attribute (RFC 1327).
     (active, dropck_eyepatch, "1.10.0", Some(34761), None),
 
-    // Allows the use of custom attributes (RFC 572).
-    (active, custom_attribute, "1.0.0", Some(29642), None),
+    // Allows using `#[structural_match]` which indicates that a type is structurally matchable.
```
Comment (oli-obk) [inline src/libsyntax/feature_gate.rs]:
```suggestion
    // Allows using `#[structural_match]` which indicates that constants of that type
    // may be used in patterns
```

File: src/libsyntax/feature_gate.rs [hunk 4]
```diff
@@ -491,7 +561,6 @@ declare_features! (
     (removed, visible_private_types, "1.0.0", None, None, None),
     (removed, unsafe_no_drop_flag, "1.0.0", None, None, None),
     // Allows using items which are missing stability attributes
-    // rustc internal
```
[the remark this answers is not in the document]
Comment (Centril) [reply src/libsyntax/feature_gate.rs]: The `// rustc internal` mechanism was pretty fragile and imo unreadable as to what it applied to (I really didn't understand what it applied to when reading the list), it looked for consecutive lines unbroken by empty lines and `Some($number)`. I replaced it with:

```
// no tracking issue START

$contents

// no tracking issue END
```

to fit with the new, clearer, and improved grouping style where:

```
    // -------------------------------------------------------------------------
    // Internal feature gates.
    // -------------------------------------------------------------------------
```
tells you what is internal.

File: src/tools/tidy/src/features.rs [hunk 5]
```diff
@@ -172,19 +172,23 @@ fn test_filen_gate(filen_underscore: &str, features: &mut Features) -> bool {
 pub fn collect_lang_features(base_src_path: &Path, bad: &mut bool) -> Features {
     let contents = t!(fs::read_to_string(base_src_path.join("libsyntax/feature_gate.rs")));
 
-    // We allow rustc-internal features to omit a tracking issue.
-    // These features must be marked with a `// rustc internal` in its own group.
```
[the remark this answers is not in the document]
Comment (Centril) [reply src/tools/tidy/src/features.rs]: (not applicable anymore with `// no tracking issue START`)

File: src/libsyntax/feature_gate.rs (hunk 1 above)
Comment (Centril) [reply src/libsyntax/feature_gate.rs]: Tidy got angry with me when I actually wrote it this way which is why I changed it ;)

File: src/libsyntax/feature_gate.rs (hunk 3 above)
Comment (Centril) [reply src/libsyntax/feature_gate.rs]: I believe `#[structural_match]` does / will do more than this; See: https://github.com/rust-lang/rfcs/blob/master/text/2000-const-generics.md#structural-equality wherein `#[structural_match]` is required for const generics; cc @varkor cc https://github.com/rust-lang/rust/issues/60286

File: src/libsyntax/feature_gate.rs [hunk 6]
```diff
@@ -108,124 +108,227 @@ macro_rules! declare_features {
...
-    //
-    // rustc internal
-    (active, prelude_import, "1.2.0", None, None),
+    // -------------------------------------------------------------------------
+    // Actual feature gates (target features).
+    // -------------------------------------------------------------------------
 
-    // Allows default type parameters to influence type inference.
-    (active, default_type_parameter_fallback, "1.3.0", Some(27336), None),
+    // FIXME: Document these and merge with the list below.
```
[the remark this answers is not in the document]
Comment (Centril) [reply src/libsyntax/feature_gate.rs]: It didn't fit with the other grouping which is a sorted list so I moved here temporarily and hopefully @gnzlbg can help me reintegrate it later.

File: src/libsyntax/feature_gate.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (oli-obk) [reply src/libsyntax/feature_gate.rs]: right, I got that. What I meant is that these were internal before and aren't anymore after your PR.

File: src/libsyntax/feature_gate.rs (hunk 1 above)
Comment (oli-obk) [reply src/libsyntax/feature_gate.rs]: ?? why would tidy complain about this?

File: src/tools/tidy/src/features.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (oli-obk) [reply src/tools/tidy/src/features.rs]: the part about allowing internal features to omit tracking issues is still applicable

File: src/libsyntax/feature_gate.rs (hunk 1 above)
Comment (Centril) [reply src/libsyntax/feature_gate.rs]: I get: `tidy error: /home/centril/programming/rust/master/src/libsyntax/feature_gate.rs:152: malformed stability attribute` cause tidy looks for lines containing `rustc_const_unstable(`.

File: src/tools/tidy/src/features.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Centril) [reply src/tools/tidy/src/features.rs]: Oh; You mean like that... How about something like:
```
    // We allow rustc-internal features to omit a tracking issue.
    // To make tidy accept omitting a tracking issue, group the list of features
    // without one inside `// no tracking issue START` and `// no tracking issue END`.
```

File: src/libsyntax/feature_gate.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Centril) [reply src/libsyntax/feature_gate.rs]: I don't follow; everything without a tracking issue is grouped inside `// no tracking issue START/END`... The list of internal features are not limited to those with no tracking issue.

File: src/libsyntax/feature_gate.rs (hunk 1 above)
Comment (Centril) [reply src/libsyntax/feature_gate.rs]: Fixed it by using a more well formed example.

File: src/tools/tidy/src/features.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Centril) [reply src/tools/tidy/src/features.rs]: Changed to ^--

File: src/libsyntax/feature_gate.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (oli-obk) [reply src/libsyntax/feature_gate.rs]: the ones below this comment have no `// no tracking issue START/END` around them. What I'm wondering is why tidy doesn't complain, but I guess that's fine for `removed` feature gates

[the remark this answers is not in the document]
Comment (Centril) [reply src/libsyntax/feature_gate.rs]: Yeah it only complains about `active` ones with no issue.

**Pull request merged**